### PR TITLE
store: add proof to ViewState response

### PR DESCRIPTION
**Pull request by blasrodri**
Introduce visited_nodes to TrieIterator which records raw bytes of all
the nodes the iterator visits.  Those are then returned as proof in
response to ViewState request.  Since the exact nodes the iterator
visited are included in the proof, it is therefore possible to verify
all the values returned in the state.

At the moment the proof is always included.

Fixes: https://github.com/near/nearcore/pull/2076

### CODE REVIEW

**Comment by blasrodri** [comment]
@mina86 I think I've broken something when merging master. Can you take a look?

File: core/store/src/trie/mod.rs [hunk 1]
```diff
@@ -397,7 +397,7 @@ impl RawTrieNode {
         out
     }
 
-    fn decode(bytes: &[u8]) -> Result<Self, std::io::Error> {
+    pub fn decode(bytes: &[u8]) -> Result<Self, std::io::Error> {
```
Comment (mina86) [inline core/store/src/trie/mod.rs:400]: I think this doesn’t need to be pub

File: core/store/src/trie/iterator.rs [hunk 2]
```diff
@@ -148,7 +170,10 @@ impl<'a> TrieIterator<'a> {
...
     /// Fetches block by its hash and adds it to the trail.
     ///
-    /// The node is stored as the last [`Crumb`] in the trail.
+    /// The node is stored as the last [`Crumb`] in the trail.  If iterator is
+    /// configured to remember all the nodes its visiting (which can be enabled
+    /// with [`Self::remember_visited_nodes`]), the node will be added to the
+    /// list.
     fn descend_into_node(&mut self, hash: &CryptoHash) -> Result<(), StorageError> {
         let node = self.trie.retrieve_node(hash)?.1;
         self.trail.push(Crumb { status: CrumbStatus::Entering, node, prefix_boundary: false });
```
Comment (mina86) [inline core/store/src/trie/iterator.rs:179]: This should be

```suggestion
        let (bytes, node) = self.trie.retrieve_node(hash)?;
        if let Some(ref mut visited) = self.visited_nodes {
            visited.push(bytes);
        }
        self.trail.push(Crumb { status: CrumbStatus::Entering, node, prefix_boundary: false });
```

**Comment by mina86** [comment]
> cutting one element (the last) element to the proof works (!)

Yes, that’s expected.  Previously iterator would fetch an unnecessary node which ended up included in the proof.  With the seek_prefix change this no longer happens and the proof is one node shorter.

File: runtime/near-vm-runner/src/tests/fuzzers.rs [hunk 3]
```diff
@@ -2,7 +2,6 @@ use arbitrary::Arbitrary;
 use bolero::check;
 use core::fmt;
 use near_primitives::contract::ContractCode;
-use near_primitives::hash::CryptoHash;
```
Comment (mina86) [inline runtime/near-vm-runner/src/tests/fuzzers.rs:5]: Bogus change?

**Review by mina86 — APPROVED** [review]
Great, thanks!

Sorry it took so long, but it turned out to be more convoluted than it appeared at first sight.  Especially since it is part of public API and we had the whole borsh red herring at the beginning.  But we’re finally here! \o/

**Comment by blasrodri** [comment]
> Great, thanks!
> 
> Sorry it took so long, but it turned out to be more convoluted than it appeared at first sight. Especially since it is part of public API and we had the whole borsh red herring at the beginning. But we’re finally here! \o/

I'm going to cry! Thanks for your support @mina86

**Pull request merged**